Add `correct_misspelled_values()` function

joshwlambert opened the pull request:
This is a ~~draft~~ PR to sketch a possible new function for the {cleanepi} package: `clean_spelling_mistakes()`. This function uses approximate string matching to a list of user-specified words (i.e. a dictionary) and the spelling mistakes that are within the `max.distance` argument (from `agrepl()`) are matched and replaced by the word in the dictionary. 

~~This is just a quick draft of a function so~~ Any comments and suggestions are welcomed. I've tried to stick to the coding style used elsewhere in {cleanepi} but where it differs I'll happily make updates.

joshwlambert commented:
Moving this from a draft PR to ready for review as I'm happy with the first implementation. It definitely needs more testing as I don't fully understand the `max.distance` argument in `agrepl()` which is a crucial component of this function. 

@Karim-Mane please can you test it out and let me know what you think?

Once we've established the functionality I can write some more unit tests before we merge this PR.

@Karim-Mane I'm happy for you to push to this PR with edits or I can make updates if you'd prefer to leave comments, whatever works best for you.

Bisaloo commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,98 @@
+#' @inheritParams clean_data
+#' @param wordlist A `character` vector or list of words to match to cells in
+#' `data`.
+#' @inheritParams base::agrep
```
Should these be passed as `...` for future-proofing the potential addition of new arguments to `agrep()`?

joshwlambert replied:
Resolved in 142d1251b3b718d59f4e54f8279fb2adab12c20a.

Bisaloo commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,98 @@
+  # remove correct spelling from matches
+  idx[df_col == word] <- FALSE
+  if (any(idx)) {
+    # only show user menu when interactive
+    if (rlang::is_interactive() && confirm) {
```
Instead of having this under the hood, my initial preference would be to have `rlang::is_interactive()` as the default for `confirm` but to let the user override it.

Could you double check what other pkgs with a TUI (e.g., usethis) are doing here please?

joshwlambert replied:
Resolved in b27bd2815f8169b471f606388ba13286f880139e.

Bisaloo commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,98 @@
+  wordlist <- unlist(wordlist)
+
+  for (col in seq_len(ncol(data))) {
+    for (word in seq_along(wordlist)) {
```
By using `adist()` instead a `agrep()`, you can vectorize it over `word` and avoid a loop here.

``` r
adist(
  c("confirmed", "confermed", "probable", "susspected"),
  c("confirmed", "probable", "suspected", "died", "recovered")
)
#>      [,1] [,2] [,3] [,4] [,5]
#> [1,]    0    8    7    6    6
#> [2,]    1    8    6    7    5
#> [3,]    8    0    8    8    8
#> [4,]    8    9    1    8    7
```

<sup>Created on 2025-03-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

It will also force the explicit choice of what happens for ties in potential fixes.

The current version will prioritize the last tie but is it what we want?

joshwlambert replied:
Resolved in 8402a6d820d64a7c1ac79d2c404bb19239f43766.

I'd appreciate this change being double checked to make sure the functionality is as we expect.

joshwlambert commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,98 @@
+#' word.
+#'
+#' @inheritParams clean_data
+#' @param wordlist A `character` vector or list of words to match to cells in
+#' `data`.
```
Happy to change to just accepting a vector. I think the reason I did this and then `unlist()` in the function is I thought some users might assume its a list because the argument is called `wordlist`.

joshwlambert replied:
Resolved in 21352b5d931eda059a76ee8f7bace6db67acc9ee.

joshwlambert commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,98 @@
+  for (col in seq_len(ncol(data))) {
+    for (word in seq_along(wordlist)) {
+      # only check and fix char columns
+      if (is.character(data[, col])) {
```
Resolved in 5e44cb1a7a377912d48c08e3e5cd1e6be47627cc.

To be honest, I think this adds friction for users that want to do a sweep across the `data.frame` and apply spelling fixes. I've not tried out all the other functions in {cleanepi} but I would say it also provides friction for them as well. From what I can tell there is no easy way for the user to specify to target all columns. 

@Karim-Mane can we think about having an extra option, like using `NULL` or `"all"` as the default to easily allow the user to apply the function to every column?

Bisaloo reviewed (commented):
As far as I can tell, there is no explicit handling of ties. The latest best match will be selected. This is ok as a design choice but if it is an explicit choice, it would be good to explicitly document it.

Additionally, correctly spelled words can be sometimes be corrected when they shouldn't.

See the following example (admittedly, an edge carefully picked to be as annoying as possible):

```
df <- data.frame(
  case_type = c("confirmed", "confermed", "probable", "susspected"),
  outcome = c("died", "recoverd", "did", "recovered")
)
clean_spelling_mistakes(
  data = df,
  target_columns = c("case_type", "outcome"),
  wordlist = c("confirmed", "probable", "suspected", "died", "recovered", "diad"),
  confirm = FALSE
)
```

Karim-Mane commented on DESCRIPTION:
```diff
@@ -52,7 +52,8 @@ Imports:
     rlang,
-    tibble
+    tibble,
+    utils
```
No need to specify {utils} as a dependency as it is part of base R. You can double check by running running R CMD CHECK without after removing it.

Karim-Mane commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,110 @@
+#' Clean spelling mistakes detected with approximate string matching and
```
```suggestion
#' Correct misspelled values by using approximate string matching techniques to
#' compare them against the expected values.
#'
#' @details
#' When used interactively (see [interactive()]) the user is presented a menu
#' to ensure that the words detected using approximate string matching are not
#' false positives and the user can decided whether to proceed with the
#' spelling corrections. In non-interactive sessions all misspelled values are
#' replaced by their closest values within the provided vector of expected
#' values.
#'
#' @inheritParams clean_data
#' @param target_columns A \code{<vector>} of the target column names. When the
#'    input data is a \code{<linelist>} object, this parameter can be set to
#'    \code{linelist_tags} to apply the fuzzy matching exclusively to the
#'    tagged columns.
#' @param wordlist A \code{<vector>} of characters with the words to match to
#'    the detected misspelled values.
#' @param confirm A `logical` that determines whether to show the user a menu of
#'    spelling corrections. If `TRUE` and using \R interactively then the user
#'    will have the option to review the proposed spelling corrections. This
#'    argument is useful for turning off the [menu()] when
#'    [rlang::is_interactive()] returns `TRUE` but not wanting to prompt the
#'    user e.g. `devtools::run_examples()`.
```

joshwlambert replied:
This could not be committed due to conflicting changes, applied in cebaa7d0e59a559c5295ab44c41de69365a2313f.

Karim-Mane commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,110 @@
+#' e.g. `devtools::run_examples()`).
+#' @param ... [dots] Extra arguments to pass to [adist()].
+#'
+#' @return The cleaned input date according to the user-specified `wordlist`.
```
```suggestion
#' @return The cleaned input data according to the user-specified `wordlist`.
```

Karim-Mane commented on R/clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,110 @@
+#'   wordlist = c("confirmed", "probable", "suspected", "died", "recovered"),
+#'   confirm = FALSE
+#' )
+clean_spelling_mistakes <- function(data,
```
```suggestion
correct_misspelled_values <- function(data,
```

Karim-Mane reviewed (changes requested):
@joshwlambert - this looks good to me. I made suggestions around:

* changing the function name from `clean_spelling_mistakes()` to `correct_misspelled_values()` to comply with the naming convention used in {cleanepi}. See the package design vignette to see how the usage of `clean_` and `correct_` prefixes are defined.
* using the same function documentation style as in the other {cleanepi} functions.

joshwlambert commented:
Hi @avallecam, this has been on the back-burner for a while. I'll try and commit the changes for @Karim-Mane to review next week if the feature this PR adds is desired.

joshwlambert commented on tests/testthat/test-clean_spelling_mistakes.R:
```diff
@@ -0,0 +1,20 @@
+    clean_df$outcome,
+    c("died", "recovered", "died", "recovered")
+  )
+})
```
I've implemented this feature of allowing the user to choose if there are multiple equal matches when they're using the package interactively and set `confirm = TRUE` in 89bb9d99a81584dd172dd236565e1c6ee4600131.

joshwlambert commented:
@Karim-Mane I've now pushed all the changes and have hopefully addressed the points raised in this PR. I'll hand this over to you to make any modification you wish and then hopefully this can be merged. 

If you have any questions about the functionality please let me know. The new code added could ideally be tidied up a bit, but I don't have the time to dedicate to this now, feel free to tidy it up in this PR, it can easily be done in a future PR.
